modules/apps/27-interchain-accounts/controller/keeper/msg_server_test.go

```diff
@@ -0,0 +1,88 @@
+package keeper_test
+
+import (
+	sdktypes "github.com/cosmos/cosmos-sdk/types"
+
+	icatypes "github.com/cosmos/ibc-go/v5/modules/apps/27-interchain-accounts/controller/types"
+	channeltypes "github.com/cosmos/ibc-go/v5/modules/core/04-channel/types"
+	host "github.com/cosmos/ibc-go/v5/modules/core/24-host"
+	ibctesting "github.com/cosmos/ibc-go/v5/testing"
+)
+
+func (suite *KeeperTestSuite) TestRegisterAccount() {
+	var (
+		msg               *icatypes.MsgRegisterAccount
+		expectedChannelID = "channel-0"
+	)
+
+	testCases := []struct {
+		name     string
+		expPass  bool
+		malleate func()
+	}{
+		{
+			"success",
+			true,
+			func() {},
+		},
+		{
+			"invalid connection id",
+			false,
+			func() {
+				msg.ConnectionId = "connection-100"
+			},
+		},
+		{
+			"non-empty owner address is valid",
+			true,
+			func() {
+				msg.Owner = "<invalid-owner>"
+			},
+		},
+		{
+			"empty address invalid",
+			false,
+			func() {
+				msg.Owner = ""
+			},
+		},
+		{
+			"port is already bound for owner but capability is claimed by another module",
+			false,
+			func() {
+				capability := suite.chainA.GetSimApp().IBCKeeper.PortKeeper.BindPort(suite.chainA.GetContext(), TestPortID)
+				err := suite.chainA.GetSimApp().TransferKeeper.ClaimCapability(suite.chainA.GetContext(), capability, host.PortPath(TestPortID))
+				suite.Require().NoError(err)
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		suite.SetupTest()
+
+		path := NewICAPath(suite.chainA, suite.chainB)
+		suite.coordinator.SetupConnections(path)
+
+		msg = icatypes.NewMsgRegisterAccount(ibctesting.FirstConnectionID, ibctesting.TestAccAddress, "")
+
+		tc.malleate()
+
+		cacheCtx, writeCache := suite.chainA.GetContext().CacheContext()
+		res, err := suite.chainA.GetSimApp().ICAControllerKeeper.RegisterAccount(sdktypes.WrapSDKContext(cacheCtx), msg)
```

Comment:
I don't think we need to cache the context? `suite.SetupTest()` will give a fresh testing state for each test case

Reply:
the purpose of the context is to verify that the correct events were emitted, is there a way of doing this without using a cached context? (my understanding was this was creating a new cache for each test run)